src/test/app/Offer_test.cpp

```diff
@@ -795,22 +850,831 @@ class Offer_test : public beast::unit_test::suite
             owners ("eve", 1));
     }
 
+    void
+    testcaseSelfCross()
+    {
+        testcase ("Self-crossing");
+
+        using namespace jtx;
+
+        auto const gw = Account {"gateway"};
+        auto const USD = gw["USD"];
+        auto const BTC = gw["BTC"];
+
+        Env env {*this};
+        env.fund (XRP (10000), gw);
+        env (offer (gw, BTC (500), USD (100)));
+        env (offer (gw, USD (100), BTC (500)));
+        env.close ();
```

Comment:
Hmmm.  This shows that self crossing is possible, but it doesn't look at the results.  It would be nice if this test looked at the state of gw's offers at the end.

BTW, things may not be what you expect after the call to `close`.  Execution of transactions between one `close` and the next is not based on the order of arrival, it's based on a numeric sort of the transaction hashes.  So, if you want to call `close` in unit tests you want to call it often.  In particular, `fund` might not self-call `close` (I'm not certain).  If `fund` doesn't self-call `close`, then either or both of the offers might get created before gw is funded, in which case those offers will not end up in the closed ledger.

Since learning and internalizing this, my tendency with unit tests has been to call `close` often and only check the state of a closed ledger.  Because, at the end of the day, what really matters to us is what got into a closed ledger.  But that's just been my approach.  I've seen lots of tests that never call `close` at all.  Your choice.